pkg/controller/deployment/deployment_controller.go

```diff
@@ -816,30 +834,131 @@ func (dc *DeploymentController) reconcileOldRCs(allRCs []*api.ReplicationControl
+
+	// Check if we can scale down. We can scale down in the following 2 cases:
+	// * Some old rcs have unhealthy replicas, we could safely scale down those unhealthy replicas since that won't further
+	//  increase unavailability.
+	// * New rc has scaled up and it's replicas becomes ready, then we can scale down old rcs in a further step.
+	//
+	// maxScaledDown := allPodsCount - minAvailable - newRCPodsUnavailable
+	// take into account not only maxUnavailable and any surge pods that have been created, but also unavailable pods from
+	// the newRC, so that the unavailable pods from the newRC would not make us scale down old RCs in a further step(that will
+	// increase unavailability).
+	//
+	// Concrete example:
+	//
+	// * 10 replicas
+	// * 2 maxUnavailable (absolute number, not percent)
+	// * 3 maxSurge (absolute number, not percent)
+	//
+	// case 1:
+	// * Deployment is updated, newRC is created with 3 replicas, oldRC is scaled down to 8, and newRC is scaled up to 5.
+	// * The new RC pods crashloop and never become available.
+	// * allPodsCount is 13. minAvailable is 8. newRCPodsUnavailable is 5.
+	// * A node fails and causes one of the oldRC pods to become unavailable. However, 13 - 8 - 5 = 0, so the oldRC won't be scaled down.
+	// * The user notices the crashloop and does kubectl rollout undo to rollback.
+	// * newRCPodsUnavailable is 1, since we rolled back to the good RC, so maxScaledDown = 13 - 8 - 1 = 4. 4 of the crashlooping pods will be scaled down.
+	// * The total number of pods will then be 9 and the newRC can be scaled up to 10.
+	//
+	// case 2:
+	// Same example, but pushing a new pod template instead of rolling back (aka "roll over"):
+	// * The new RC created must start with 0 replicas because allPodsCount is already at 13.
+	// * However, newRCPodsUnavailable would also be 0, so the 2 old RCs could be scaled down by 5 (13 - 8 - 0), which would then
+	// allow the new RC to be scaled up by 5.
 	minAvailable := deployment.Spec.Replicas - maxUnavailable
-	minReadySeconds := deployment.Spec.MinReadySeconds
-	// Check the expectations of deployment before counting available pods
-	dKey, err := controller.KeyFunc(&deployment)
-	if err != nil {
-		return false, fmt.Errorf("Couldn't get key for deployment %#v: %v", deployment, err)
+	newRCPodsUnavailable := newRC.Spec.Replicas - newRCReadyPodCount
+	maxScaledDown := allPodsCount - minAvailable - newRCPodsUnavailable
+	if maxScaledDown <= 0 {
+		return false, nil
 	}
-	if expectationsCheck && !dc.podExpectations.SatisfiedExpectations(dKey) {
-		glog.V(4).Infof("Pod expectations not met yet before reconciling old RCs\n")
+
+	// Clean up unhealthy replicas first, otherwise unhealthy replicas will block deployment
+	// and cause timeout. See https://github.com/kubernetes/kubernetes/issues/16737
+	cleanupCount, err := dc.cleanupUnhealthyReplicas(oldRCs, deployment, maxScaledDown)
+	if err != nil {
 		return false, nil
 	}
+
+	totalScaledDown := cleanupCount
```

Comment:
The cleaned-up pods shouldn't have reduced the number available, so cleanupCount shouldn't be relevant. The idea was to remove unproductive pods from oldRCs so that the rollout could make forward progress. But not excessively rapid forward progress, which is why we computed maxScaledDown.

At this point, we want to scale down if and only if there is a surplus of available pods (i.e., more than minAvailable).

Since scaleDownOldRCsForRollingUpdate computes minAvailable, I don't think we need to pass it another scaling limit.